MPS: Connect to GameLift with Player Session Data (#370)
o3de/o3de-multiplayersample: README_GAMELIFT.md

Reviewer: Consider moving this into the Documentation folder and linking from main README

Helps keep things more compact.

This README covers testing and running MultiplayerSample with Amazon GameLift.

## Running with [GameLift](https://docs.aws.amazon.com/gamelift/index.html)

### Setup
1. Install the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html), if you don't already have it. You will need it to create a game session with GameLiftLocal.

> 1. Enable the "AWSGameLift" and "MPSGameLift" gem by adding them to MultiplayerSample/Gem/Code/enabled_gems.cmake

Reviewer: ```suggestion
1. Enable the "AWSGameLift" and "MPSGameLift" gems by adding them to `MultiplayerSample/Gem/Code/enabled_gems.cmake`
```

> 1. Build the server and game launchers for MultiplayerSample as normal, per [top-level README](/README.md).

Reviewer: ```suggestion
1. Build the server and game launchers for MultiplayerSample as normal, per the [README](/README.md).
```

1. Work in progress (WiP) step: Add your AWS region to Config/default_aws_resource_mappings.json (example: "Region": "us-west-2")
    a. Currently needed otherwise when the client initializes GameLift there will be an error about not having a region. 
    b. This step will be removed once we properly parse the game-session data which contains the fleet-id, region-id, etc  

## Build Server for Windows (WiP)

> 1. WiP Step: Build a profile pak server/game build

Reviewer: Can link to documentation here: https://github.com/o3de/o3de-multiplayersample/pull/321

    a. This step should be replaced by building a proper release build
1. Build Monolithic Server
    a. cmake -B build\windows_mono -S . -G "Visual Studio 16" -DLY_MONOLITHIC_GAME=1 -DALLOW_SETTINGS_REGISTRY_DEVELOPMENT_OVERRIDES=0
    b. cmake --build build\windows_mono --target MultiplayerSample.GameLauncher MultiplayerSample.ServerLauncher MultiplayerSample.UnifiedLauncher --config profile -- /m /nologo
1. Bundle Content
    a. build\windows\bin\profile\AssetBundler.exe
    b. Follow steps for "Create a bundle for game assets" and "Create a bundle for engine assets" and "Add bundles to the release game layout" here: https://www.o3de.org/docs/user-guide/packaging/asset-bundler/bundle-assets-for-release/

The "default seed lists" choice should choose all but 4 seed lists to make the engine_pc.pak
The other 4 seed lists should all get selected to make the game_pc.pak
It's important to make sure that the bootstrap.game.profile.setreg file has been added to one of the seed files. (also add debug if you want to support debug builds)
1. Create the Launcher Zip file

>    Use the following .bat file or equivalent copy steps to create a directory with the launchers in it:

Reviewer: nit: Windows specific, so may want to highlight

   ```sh

>     rem Use this by calling 'make_release D:\my\output\GameLiftPackageWindows' to make a release directory

Reviewer: ```suggestion
    rem Use this by calling 'make_release C:\path\to\my\GameLiftPackageWindows' to make a release directory
```

    mkdir %1
    mkdir %1\Cache
    mkdir %1\Cache\pc
    mkdir %1\Gems
    mkdir %1\Gems\AWSCore
    
    rem Copy the pak files

>     copy D:\github\o3de-multiplayersample\AssetBundling\Bundles\*.pak %1\Cache\pc

Reviewer: ```suggestion
    copy C:\path\to\o3de-multiplayersample\AssetBundling\Bundles\*.pak %1\Cache\pc
```

    
    rem Copy the executables and DLLs

>     copy D:\github\o3de-multiplayersample\build\windows_mono\bin\profile\*.* %1

Reviewer: More D: paths

Author: fixed to just use relative path. users can change it, or just place the script in their root folder. future upgrade could be to ask user to provide the path as input.

    
    rem Copy the AWSCore files
    copy D:\github\o3de-multiplayersample\build\windows_mono\bin\profile\Gems\AWSCore\*.* %1\Gems\AWSCore
    
    rem Copy the launch_client / launch_server files
    copy D:\github\o3de-multiplayersample\launch_*.* %1
    ```

## Prepare for GameLift
### Upload the build to GameLift
Note: Builds are tied to Fleets; you may want to delete the existing build and fleet via the AWS Gamelift dashboard just so you don't accidentally reference old builds or old fleets in future steps. 
`
aws gamelift upload-build --operating-system WINDOWS_2012 --build-root .\GameLiftPackageWindows\ --name MultiplayerSample --build-version v1.0 --region us-west-2
`
This returns the build id which will be used in the next steps. Example Build ID: build-f447776f-b4ba-4f1e-99cf-09d49f114942

### Create Fleet
After running this command it'll take about an hour for the fleet to activate. Check the status on the GameLift dashboard. 

`

> aws gamelift create-fleet --region us-west-2 --name GlLyTest2F --ec2-instance-type c5.large --fleet-type ON_DEMAND --build-id build-f447776f-b4ba-4f1e-99cf-09d49f114942 --runtime-configuration "GameSessionActivationTimeoutSeconds=300, MaxConcurrentGameSessionActivations=2, ServerProcesses=[{LaunchPath=C:\game\MultiplayerSample.ServerLauncher.exe, Parameters= --sv_gameLiftEnabled=true --sv_dedicated_host_onstartup=false --loadlevel=NewStarbase, ConcurrentExecutions=10}]" --ec2-inbound-permissions "FromPort=33435,ToPort=33435,IpRange=0.0.0.0/0,Protocol=UDP" "FromPort=33235,ToPort=33235,IpRange=0.0.0.0/0,Protocol=UDP"

Reviewer: ```suggestion
aws gamelift create-fleet --region us-west-2 --name GlLyTest2F --ec2-instance-type c5.large --fleet-type ON_DEMAND --build-id build-f447776f-b4ba-4f1e-99cf-09d49f114942 --runtime-configuration "GameSessionActivationTimeoutSeconds=300, MaxConcurrentGameSessionActivations=1, ServerProcesses=[{LaunchPath=C:\game\MultiplayerSample.ServerLauncher.exe, Parameters= --sv_gameLiftEnabled=true --sv_dedicated_host_onstartup=false --loadlevel=NewStarbase, ConcurrentExecutions=1}]" --ec2-inbound-permissions "FromPort=33435,ToPort=33435,IpRange=0.0.0.0/0,Protocol=UDP" "FromPort=33235,ToPort=33235,IpRange=0.0.0.0/0,Protocol=UDP"
```

Unclear why you are opening two UDP ports and why you are launching 10 processes

Author: Fixed! This burnt me as well, readme started as a copy from old documentation. I didn't have the correct ports open so the client could never connect.

`

Record the fleet-id for the next step.

### Create and Join Game Session
`

> aws gamelift create-game-session --region us-west-2 --fleet-id fleet-6ae4d52f-4a6b-4591-99da-122d37922490 --name foogamesession1 --maximum-player-session-count 3 --game-properties Key=sv_map,Value=MultiplayerSample

Reviewer: ```suggestion
aws gamelift create-game-session --region us-west-2 --fleet-id <fleet-id from create fleet> --name foogamesession1 --maximum-player-session-count 3 --game-properties Key=sv_map,Value=MultiplayerSample
```

Author: fixed

`

`
aws gamelift create-player-session --region us-west-2 --game-session-id arn:aws:gamelift:us-west-2::gamesession/fleet-6ae4d52f-4a6b-4591-99da-122d37922490/gsess-f2c14a63-e5fe-462f-9f8e-6be95b9dda26 --player-id Player1
`

Take the output of create-player-session and reformat the values to the custom session token like below, use this in the game immediately because it expires after 60 seconds.

> Launch the game client with:

Reviewer: Don't you:
a) Want to launch client first and get to connection screen
b) then make player session

Also you are launching the client to talk to GameLift local? which seems confusing having just made a fleet on the GameLift service above?

Author: Good call about opening GameLauncher before creating the player session because the player session times out after 1 minute. Fixed!

_Also you are launching the client to talk to GameLift local? which seems confusing having just made a fleet on the GameLift service above?_
This doc is for GameLift not GameLift local. Should be better now?
